### PR TITLE
Return a PluginError instead of a plain error

### LINKED ISSUE #694: CreateOrUpdate should return type 'PluginError' instead of 'error'
https://github.com/openshift/openshift-azure/blob/master/pkg/plugin/plugin.go#L103 should return type 'PluginError' instead of 'error'. This is needed so that RP/service: Can expect error to be in an opinionated format and custom passing will be done on `Err` field of 'PluginError' by the RP.

cc: @asalkeld @jim-minter

### PR DESCRIPTION
**Pull request by thekad**
Closes #694 

Simple change to return an opinionated structure instead of a plain error

### CODE REVIEW

File: pkg/fakerp/fakerp.go [hunk 1]
```diff
@@ -145,9 +145,9 @@ func CreateOrUpdate(ctx context.Context, oc *v20180930preview.OpenShiftManagedCl
 		return nil
 	}
 
-	err = p.CreateOrUpdate(ctx, cs, azuretemplate, oldCs != nil, deployer)
-	if err != nil {
-		return nil, err
+	perr := p.CreateOrUpdate(ctx, cs, azuretemplate, oldCs != nil, deployer)
```
Comment (thekad) [inline pkg/fakerp/fakerp.go]: I feel like the fakerp should also return a `PluginError` here, thoughts?

File: pkg/plugin/plugin.go [hunk 2]
```diff
@@ -110,15 +110,16 @@ func (p *plugin) CreateOrUpdate(ctx context.Context, cs *api.OpenShiftManagedClu
...
 	if err != nil {
-		return err
+		return api.PluginError{Err: err, Step: api.PluginStepDeploy}
 	}
 
 	// Wait for infrastructure services to be healthy
 	err = p.clusterUpgrader.WaitForInfraServices(ctx, cs)
 	if err != nil {
-		return err
+		return api.PluginError{Err: err, Step: api.PluginStepWaitForInfraDeployments}
```
Comment (thekad) [inline pkg/plugin/plugin.go]: not sure about this step here, there are several `WaitForInfra` steps...

File: pkg/plugin/plugin.go [hunk 3]
```diff
@@ -110,15 +110,16 @@ func (p *plugin) CreateOrUpdate(ctx context.Context, cs *api.OpenShiftManagedClu
 		err = p.clusterUpgrader.Deploy(ctx, cs, azuretemplate, deployFn)
 	}
 	if err != nil {
-		return err
+		return api.PluginError{Err: err, Step: api.PluginStepDeploy}
```
Comment (thekad) [inline pkg/plugin/plugin.go]: should I return a diff step on deploy vs update?

**Comment by asalkeld** [comment]
@thekad I think you only need to change the return type in plugin and in the upgrader.
https://github.com/openshift/openshift-azure/blob/master/pkg/cluster/types.go#L23-L28
you shouldn't have to add new PluginErrors as the upgrader already returns the required errors.

Although, IMHO returning `error` is the correct approach.

File: pkg/api/plugin.go [hunk 4]
```diff
@@ -92,5 +92,5 @@ type Plugin interface {
 
 	// CreateOrUpdate either deploys or runs the update depending on the isUpdate argument
 	// this will call the deployer.
-	CreateOrUpdate(ctx context.Context, cs *OpenShiftManagedCluster, azuretemplate map[string]interface{}, isUpdate bool, deployer DeployFn) error
+	CreateOrUpdate(ctx context.Context, cs *OpenShiftManagedCluster, azuretemplate map[string]interface{}, isUpdate bool, deployer DeployFn) PluginError
```
Comment (jim-minter) [inline pkg/api/plugin.go]: this should be `*PluginError`, so that it can also be nil.

File: pkg/fakerp/fakerp.go (hunk 1 above)
Comment (jim-minter) [reply pkg/fakerp/fakerp.go]: once CreateOrUpdate returns a `*PluginError`, this entire hunk can be reverted

File: pkg/plugin/plugin.go (hunk 3 above)
Comment (jim-minter) [reply pkg/plugin/plugin.go]: this isn't right.  Update and Deploy already effectively return *PluginError, it's just that their signatures are wrong.  Change their signatures.  Ditto WaitForInfraServices, etc.

File: pkg/cluster/types.go [hunk 5]
```diff
@@ -22,10 +22,10 @@ import (
 // Upgrader is the public interface to the upgrade module used by the plugin.
 type Upgrader interface {
 	Initialize(ctx context.Context, cs *api.OpenShiftManagedCluster) error
-	Deploy(ctx context.Context, cs *api.OpenShiftManagedCluster, azuretemplate map[string]interface{}, deployFn api.DeployFn) error
```
Comment (jim-minter) [inline pkg/cluster/types.go]: the fact that Initialize() above is not changed implies to me that it should not be part of this interface.  Remove it from the interface, and convert its implementation to `initialize()` (unexported).

**Comment by jim-minter** [comment]
one nit then lgtm

File: pkg/cluster/initialize.go [hunk 6]
```diff
@@ -9,7 +9,7 @@ import (
 	"github.com/openshift/openshift-azure/pkg/util/azureclient/storage"
 )
 
-func (u *simpleUpgrader) Initialize(ctx context.Context, cs *api.OpenShiftManagedCluster) error {
+func initialize(u *simpleUpgrader, ctx context.Context, cs *api.OpenShiftManagedCluster) error {
```
Comment (jim-minter) [inline pkg/cluster/initialize.go]: no: should be `func (u *simpleUpgrader) initialize`.

**Comment by kargakis** [comment]
```
pkg/cluster/ready.go:159:4: cannot use err (type error) as type *"github.com/openshift/openshift-azure/pkg/api".PluginError in return argument: need type assertion
```
/lgtm cancel

**Comment by thekad** [comment]
This is odd:

```
[kad@kenpachi openshift-azure (issues/694)]$ make clean build
rm -f azure-reader.log coverage.out end-user.log e2e.test logbridge sync
go generate ./...
bindata.go
bindata.go
go build ./...
[kad@kenpachi openshift-azure (issues/694)]$ 
```

The build step obviously works on my machine, furthermore, there is no `err` variable in the location stated by the logs... What am I missing here?

**Pull request merged**